### ISSUE TITLE
How to upgrade nvidia-driver node by node?

### ISSUE THREAD

**Issue by JuHyung-Son** [NVIDIA/gpu-operator#393]
I’m planning to upgrade nvidia-driver version node by node to make some pods alive during update.

Is there any way to upgrade drivers node by node?

**Comment by faiq** [comment]
The new version of the operator v1.11.1 of the operator: https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/release-notes.html#improvements 

Where you can define `driver.rollingUpdate.maxUnavailable` parameter to specify maximum nodes for simultaneous driver upgrades. Default is 1. 

Let me know how that goes for you.

**Comment by shivamerla** [comment]
@faiq Yes, if you are using v1.11.1 then by default driver upgrades will happen node-by-node. It was supported earlier but due to a bug(missing startupProbe), driver-containers upgraded in parallel on multiple nodes simultaneously.

Also, if you mean to upgrade nodes in specific order, that is not possible with how Daemonset Rolling Upgrades happen in general. One option to achieve this specific order is to edit the nvidia-driver Daemonset and change from

```
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 1
```

to

```
spec:
  updateStrategy:
    type: OnDelete
```

With this change, upgrades are only triggered when existing Pods are manually deleted and you can control the pod deletion order here.

**Comment by faiq** [comment]
@shivamerla is it possible to expand the CRD to expose the update strategy and perhaps for other critical as well such as the toolkit

**Comment by shivamerla** [comment]
Yes, that will be useful. Will track this change for future releases.

**Comment by faiq** [comment]
@shivamerla should I create a new issue to track progress in? It would be useful for my team track the progress of the issue as well.

**Comment by shivamerla** [comment]
@faiq I will mark this one as enhancement and we can keep it open. I have created an internal one for changes.

**Comment by shivamerla** [comment]
Closing this as fixed with [this](https://github.com/NVIDIA/gpu-operator/commit/93810f4b45ef26948a703a61024249f4dd48d2b1) commit. This will be available in next patch `v22.9.1`

**Comment by shivamerla** [comment]
Please refer to more controls with driver upgrades [here](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/gpu-driver-upgrades.html)
